File: day1_part2_2018.rb

frequencies = File.readlines('input.txt').map(&:to_i)
puts "Part One: #{frequencies.sum}"

current_frequency = 0
seen_frequencies = {0 => true}
duplicate_frequency = nil

loop do
  frequencies.each do |freq|
    current_frequency += freq
    if seen_frequencies[current_frequency]
      duplicate_frequency = current_frequency
      break
    end
    seen_frequencies[current_frequency] = true
  end
  break if duplicate_frequency
end

puts "Part Two: #{duplicate_frequency}"

File: objectivec/day1_part2_2018.m
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSError *error;
        NSString *input = [NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:&error];
        NSArray *frequencyChanges = [input componentsSeparatedByString:@"\n"];
        
        // Part One
        int resultingFrequency = 0;
        for (NSString *change in frequencyChanges) {
            resultingFrequency += [change intValue];
        }
        printf("Part One - Resulting Frequency: %d\n", resultingFrequency);
        
        // Part Two
        NSMutableSet *seenFrequencies = [NSMutableSet set];
        int currentFrequency = 0;
        BOOL foundDuplicate = NO;
        
        while (!foundDuplicate) {
            for (NSString *change in frequencyChanges) {
                currentFrequency += [change intValue];
                if ([seenFrequencies containsObject:@(currentFrequency)]) {
                    printf("Part Two - First Duplicate Frequency: %d\n", currentFrequency);
                    foundDuplicate = YES;
                    break;
                }
                [seenFrequencies addObject:@(currentFrequency)];
            }
        }
    }
    return 0;
}